Feat: Adapt Backend to Ethereum-style encoding, enabling cross-chain channels

iljabvh opened the pull request:
This PR implements the following feature:

* ABI-Encoding, signing and verifying the ChannelState object such that it is compatible to Ethereum Payment Channels.
* Modfifying the Params struct to yield a ChannelID that is compatible to the ones from the Ethereum backend and contracts.

Both features allow users to participate in Payment Channels based on blockchain states on the Ethereum and CKB blockchains. The effectively allows us to use cross-chain Perun payment channels between CKB and Ethereum.

This backend work with the corresponding [contracts](https://github.com/perun-network/perun-ckb-contract/pull/19).

iljabvh committed:
Feat(channel) WIP: Add Ethereum-style encoding for cross-chain channels

iljabvh committed:
Update submodule devnet/contract to the branch ckb-eth-cross from the perun-ckb-contract repository

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
Feat(client, wallet) WIP: Include EthAddress into Participant struct, assuming separate CKB_PubKey != Eth_Pubkey

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
Update submodule devnet/contract to commit dd3e7bf4dea35ecaaf673fe6a017180e613331ec: No EthAddress in Participant anymore

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
Feat(channel, go.mod, go.sum): ABI-Encode Channel Parameters and State to calculate a cross-channel-unique ChannelID. Include cross-chain Asset, compatible with both CKB and Ethereum.

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
* Feat(wallet): Use ethereum/go/crypto for signature generation and verification.

* Test(wallet): Include TestDecredAndGethSignatures, verifying that ECDSA signatures used from the decred SDK and the go-ethereum SDK are equivalent.

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

sophia1ch committed:
test: should be reverted: remove offckb

Signed-off-by: Sophia Koehler <sophia@perun.network>

iljabvh committed:
Chore(devnet): Rebase to adjust-for-perunx, removing offckb usage and use fund_accounts.*

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
Fix(devnet, go.mod, go.sum): Change go.mod, go.sum, using the vanilla Participant type and change the submodule contract commit to a9ff15390cf64a1503c801d513da42e4e0ba5eb5

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
Update submodule to commit a5d6da1 with Keccak256 changes

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

iljabvh committed:
* Feat(channel): Introduce asset.NervosAsset, adapt remaining channel objects accordingly

* Chore(devnet/contract): Reference smart contract submodule to the most recent commit, f562afc6b5b2068b1d8e6427bd36afdcaafe571a

Signed-off-by: Ilja von Hoessle <ilja@perun.network>

sophia1ch commented on transaction/transaction.go:
```diff
@@ -603,7 +605,7 @@ func (ptb *PerunTransactionBuilder) addOrUpdateCKBChangeCell(amount uint64) erro
 	ptb.Outputs = append(ptb.Outputs, outputCell)
 	ptb.OutputsData = append(ptb.OutputsData, []byte{})
-
+	log.Println("ckbChangeCellIndex: ", len(ptb.Outputs)-1)
```
log statement can be removed

sophia1ch committed:
feat(channel/test): put omnilock back in
fix: adjust signuter and turn into molecule signature only at contract calls
fix: adjust ethAsset encoding

Signed-off-by: Sophia Koehler <sophia@perun.network>

sophia1ch committed:
feat: update sdk dependency and contract version to keep asset order intact

Signed-off-by: Sophia Koehler <sophia@perun.network>

sophia1ch committed:
feat: wrap client calls with retying blocks

Signed-off-by: Sophia Koehler <sophia@perun.network>

sophia1ch committed:
fix: remoce addition of required amount and cell capacity, otherwise swaps fail when already provided amount is inbetween required amount and required amount + cell capacity

Signed-off-by: Sophia Koehler <sophia@perun.network>

sophia1ch committed:
fix: add cell capacity

Signed-off-by: Sophia Koehler <sophia@perun.network>

sophia1ch committed:
fix: lengthen retry window

Signed-off-by: Sophia Koehler <sophia@perun.network>
